File: Java/872. Leaf-Similar Trees.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public boolean leafSimilar(TreeNode root1, TreeNode root2) {
        List<Integer> list1 = new ArrayList<Integer>();
        List<Integer> list2 = new ArrayList<Integer>();
        helper(root1, list1);
        helper(root2, list2);
        return list1.equals(list2);
    }
    public void helper(TreeNode root, List<Integer> list){
        if(root == null){
            return;
        }
        if (root.left == null && root.right == null){
            list.add(root.val);
            return;
        }
        helper(root.left, list);
        helper(root.right, list);
    }
}

File: C++/872. Leaf-Similar Trees.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool leafSimilar(TreeNode* root1, TreeNode* root2) {
        vector<int> vec1, vec2;
        helper(root1, vec1);
        helper(root2, vec2);
        return vec1 == vec2;
    }
    void helper(TreeNode* root, vector<int> &vec){
        if(root == NULL){
            return;
        }
        if (root->left == NULL && root->right == NULL){
            vec.push_back(root->val);
            return;
        }
        helper(root->left, vec);
        helper(root->right, vec);
    }
};
